src/lib/collectors/shared/launcher.ts

```diff
@@ -0,0 +1,135 @@
+import * as net from 'net';
+import * as path from 'path';
+
+import * as lockfile from 'lockfile';
+import { promisify } from 'util';
+
+import { BrowserInfo, ILauncher, LauncherOptions } from '../../types';
+import { debug as d } from '../../utils/debug';
+
+const lock = promisify(lockfile.lock);
+const unlock = promisify(lockfile.unlock);
+const debug: debug.IDebugger = d(__filename);
+
+// ------------------------------------------------------------------------------
+// Common
+// ------------------------------------------------------------------------------
+export abstract class Launcher implements ILauncher {
+    private pidFile: string = path.join(process.cwd(), 'cdp.pid');
+    protected port: number = 9222;
+    private retryDelay: number = 500;
+    protected options: LauncherOptions = null;
+
+    constructor(options: LauncherOptions) {
+        this.options = options;
+    }
+
+    /** If a browser is already running, it returns its pid. Otherwise return value is -1.  */
+    private getBrowserInfo(): BrowserInfo {
+        let result = {
+            pid: -1,
+            port: this.port
+        };
+
+        try {
+            result = JSON.parse(fs.readFileSync(this.pidFile, 'utf8'));
+        } catch (e) {
+            debug(`Error reading ${this.pidFile}`);
+            debug(e);
+            result = {
+                pid: -1,
+                port: this.port
+            };
+        }
+
+        if (Number.isNaN(result.pid)) {
+            return {
+                pid: -1,
+                port: this.port
+            };
+        }
+
+        try {
+            // We test if the process is still running or is a leftover:
+            // https://nodejs.org/api/process.html#process_process_kill_pid_signal
+
+            process.kill(result.pid, 0);
+        } catch (e) {
+            debug(`Process with ${result.pid} doesn't seem to be running`);
+            result = {
+                pid: -1,
+                port: this.port
+            };
+        }
+
+        return result;
+    }
+
+    /** Stores the `pid` of the given `child` into a file. */
+    private writePid(browserInfo: BrowserInfo): void {
+        fs.writeFileSync(this.pidFile, JSON.stringify({ pid: browserInfo.pid, port: browserInfo.port || this.port }, null, 4), { encoding: 'utf8' });
+    }
+
+    protected abstract async launchBrowser(url: string): Promise<BrowserInfo>;
+
+    /** Launches chrome with the given url and ready to be used with the Chrome Debugging Protocol.
+     *
+     * If the browser is a new instance it will return `true`, `false` otherwise.
+    */
+    public async launch(url: string): Promise<BrowserInfo> {
+        const cdpLock = 'cdp.lock';
+
+        try {
+            await lock(cdpLock, {
+                pollPeriod: 500,
+                retries: 20,
+                retryWait: 1000,
+                stale: 50000,
+                wait: 50000
+            });
+        } catch (e) {
+            console.error(e);
```

Comment:
We will need to refactor debug and logger. Maybe we can do something similar to what we do to debug where we pass the __filename to add more data to logger and merge everything.
@sarvaje can you open a new issue with this?